tomcat-gcp-lib/src/main/java/com/google/cloud/runtimes/tomcat/session/DatastoreStore.java

```diff
@@ -0,0 +1,228 @@
+import com.google.cloud.datastore.Entity;
+import com.google.cloud.datastore.Key;
+import com.google.cloud.datastore.KeyFactory;
+import com.google.cloud.datastore.KeyQuery;
+import com.google.cloud.datastore.Query;
+import com.google.cloud.datastore.QueryResults;
+import com.google.common.collect.Streams;
+
+import org.apache.catalina.LifecycleException;
+import org.apache.catalina.Session;
+import org.apache.catalina.session.StandardSession;
+import org.apache.catalina.session.StoreBase;
+import org.apache.juli.logging.Log;
+import org.apache.juli.logging.LogFactory;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.util.Arrays;
+
+/**
+ * This store interact with the Datastore service to persist sessions.
+ *
+ * <p>It does not make any assumptions about the manager, so it could be used
+ * by all manager implementations.</p>
+ *
+ * <p>However aggregations can be slow on the Datastore so if performance is a concern prefer using
+ * a manager implementation which is not using aggregations such as {@link DatastoreManager}</p>
+ */
+public class DatastoreStore extends StoreBase {
+
+  private static final Log log = LogFactory.getLog(DatastoreStore.class);
+
+  private Datastore datastore = null;
+  private KeyFactory keyFactory = null;
+  private KeyQuery.Builder keyQueryBuilder = null;
+
+  /**
+   * Name of the kind used in The Datastore for the session.
+   */
+  private String sessionKind = "GCloudSession";
+
+  /**
+   * Namespace to use in the Datastore.
+   */
+  private String namespace = "tomcat-gcp-persistent-session";
+
+  /**
+   * {@inheritDoc}
+   *
+   * <p>Initiate a connection to the Datastore.</p>
+   *
+   */
+  @Override
+  protected synchronized void startInternal() throws LifecycleException {
+    log.debug("Initialization of the Datastore Store");
+
+    this.datastore = DatastoreOptions.newBuilder().setNamespace(namespace).build().getService();
+    this.keyFactory = datastore.newKeyFactory().setKind(sessionKind);
+    this.keyQueryBuilder = Query.newKeyQueryBuilder().setKind(sessionKind);
+
+    super.startInternal();
+  }
+
+  /**
+   * Return the number of Sessions present in this Store.
+   *
+   * <p>The Datastore does not support counting element in a collection
+   * so all the keys are fetched and the count is computed locally.</p>
+   *
+   * <p>This method may be slow if a large number of sessions are persisted,
+   * prefer operations on individual entity rather than aggregations.</p>
+   *
+   * @return The number of sessions stored into the Datastore
+   */
+  @Override
+  public int getSize() throws IOException {
+    log.debug("Accessing sessions count, be cautious this operation can cause performance issues");
+    Query<Key> query = this.keyQueryBuilder.build();
+    QueryResults<Key> results = datastore.run(query);
+    long count = Streams.stream(results).count();
```

Comment:
That was my first though but the stat are computed daily therefore we can't rely on them. As I didn't find any reasonable way I specified that it may be slow in the Javadoc